src/openassetio-core-c/private/chandle.hpp

```diff
@@ -0,0 +1,39 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2013-2022 The Foundry Visionmongers Ltd
+#pragma once
+
+#include <openassetio/export.h>  // For OPENASSETIO_VERSION
+
+namespace openassetio {
+inline namespace OPENASSETIO_VERSION {
+/**
+ * Provides utilities for converting between C++ types and C handles.
+ */
+namespace chandle {
+/**
+ * Convert between C++ types and C opaque handles for those types.
+ *
+ * @tparam Type C++ type.
+ * @tparam Handle Opaque handle type.
+ */
+template <class Type, class Handle>
+struct Converter {
+  /**
+   * Convert a pointer to a C++ instance to a handle.
+   *
+   * @param ptr Pointer to C++ instance.
+   * @return Opaque C handle.
+   */
+  static Handle toHandle(Type* ptr) { return reinterpret_cast<Handle>(ptr); }
```

Comment:
Is this necessary? If the handle is a pointer to an opaque struct, and we have the definition of the struct in C++, then no conversion should be necessary?

> // foo.h
>   extern "C" {
>   typedef struct Foo* FooHandle;
>
>   FooHandle makeFoo();
>   void freeFoo(FooHandle h);
> }

> // foo.cpp
> #include "foo.h"
> #include <strings>
>
> // we can do all kinds of C++ things!
> class FooBase { virtual ~FooBase() {} };
> struct Foo : public FooBase {
> private:
>  std::string wibble = "Wobble";
> public:
>  Foo() {};
>  override ~Foo();
> }
>
> FooHandle makeFoo() { return new Foo; }
> void freeFoo(FooHandle h) { delete h; }

That should be fine, right?

Reply:
I think this is a misunderstanding.  In our approach, `struct Foo` would only be declared and never defined, it's just a dummy struct declaration, required so we can declare an opaque pointer type.

The actual data the `FooHandle` pointer points to would (likely) be an instance of a C++ namespaced and mangled class, which we `reinterpret_cast` to the C opaque pointer type for passing around in the C API.